### PR TITLE
Automatically fetch lambda layer version for sentinel

### PR DESCRIPTION
**Pull request by ben851**
# Summary | Résumé

This adds a script that will automatically fetch the latest lambda version for CDS sentinel forwarder. This means we will not have to manually track this (or use renovate PRs which could get out of hand since it's built so often).

I've also moved these helper scripts that fetch data for us (PCA ARN and Lambda Layer) inside of the aws/eks folder so that it's much easier to reference them at runtime.

## Related Issues | Cartes liées

Chore

## Before merging this PR

Read code suggestions left by the
[cds-ai-codereviewer](https://github.com/cds-snc/cds-ai-codereviewer/) bot. Address
valid suggestions and shortly write down reasons to not address others. To help
with the classification of the comments, please use these reactions on each of the
comments made by the AI review:

| Classification      | Reaction | Emoticon |
|---------------------|----------|----------|
| Useful              | +1       | 👍        |
| Noisy               | eyes     | 👀        |
| Hallucination       | confused | 😕        |
| Wrong but teachable | rocket   | 🚀        |
| Wrong and incorrect | -1       | 👎        |

The classifications will be extracted and summarized into an analysis of how helpful
or not the AI code review really is.

## Test instructions | Instructions pour tester la modification

TF Apply in staging works
Verify lambda layer is latest

## Release Instructions | Instructions pour le déploiement

None.

## Reviewer checklist | Liste de vérification du réviseur

* [x] This PR does not break existing functionality.
* [x] This PR does not violate GCNotify's privacy policies.
* [x] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
* [x] This PR does not significantly alter performance.
* [x] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+role_arn="arn:aws:iam::$ACCOUNT_ID:role/lambda-list-layers"
+session_name="${2:-session-$(date +%s)}"
+
+temp_creds=$(aws sts assume-role \
+      --role-arn "$role_arn" \
+      --role-session-name "$session_name" \
+      --duration-seconds 3600 \
+      --output json)
+
+export AWS_ACCESS_KEY_ID=$(echo $temp_creds | jq -r .Credentials.AccessKeyId)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:13]: but it works as it issss

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+export AWS_SECRET_ACCESS_KEY=$(echo $temp_creds | jq -r .Credentials.SecretAccessKey)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:14]: but it works as it issss

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+export AWS_SESSION_TOKEN=$(echo $temp_creds | jq -r .Credentials.SessionToken)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:15]: but it works as it issss

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+VERSION=$(aws lambda list-layer-versions \
+  --layer-name aws-sentinel-connector-layer \
+  --query "max_by(LayerVersions, &Version)" | \
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:19]: Done

File: aws/eks/sentinel.tf [hunk 5]
```diff
@@ -4,6 +4,13 @@ locals {
   blazer_log_group_arn      = "arn:aws:logs:${var.region}:${var.account_id}:log-group:blazer"
 }
 
+data "external" "get_sentinel_layer_version" {
+  # Get the Latest Sentinel Layer version
+  program = ["helper_scripts/getSentinelLayerVersion.sh", var.sentinel_sre_aws_account_id]
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/sentinel.tf:9]: We will get an error from terragrunt well before this if that variable doesn't exist.

File: aws/eks/vpn.tf [hunk 6]
```diff
@@ -94,8 +94,7 @@ resource "aws_acmpca_certificate" "client_vpn" {
 
 data "external" "get_pca_arn" {
   # Get the PCA ARN
-  # Yes this is horrific. I'm sorry.
-  program = [var.env == "production" ? "../../../../../../../../scripts/getPCAARN.sh" : "../../../../../../../scripts/getPCAARN.sh"]
+  program = ["helper_scripts/getPCAARN.sh"]
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/vpn.tf:97]: It is and it does

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  jq '.Version')
+
+if [-z "$VERSION"]; then
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh]: fixed

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  echo "Error: Could not get the latest version of the aws-sentinel-connector-layer"
+  exit 1
+fi
+
+jq --null-input \
+  --arg version "$VERSION" \
+  '{"version": $version}'
+
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh]: fixed

File: aws/eks/sentinel.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/sentinel.tf:9]: not needed

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+export AWS_ACCESS_KEY_ID=$(echo $temp_creds | jq -r .Credentials.AccessKeyId)
+export AWS_SECRET_ACCESS_KEY=$(echo $temp_creds | jq -r .Credentials.SecretAccessKey)
+export AWS_SESSION_TOKEN=$(echo $temp_creds | jq -r .Credentials.SessionToken)
+
+VERSION=$(aws lambda list-layer-versions \
+  --layer-name aws-sentinel-connector-layer \
+  --query "max_by(LayerVersions, &Version)" | \
+  jq '.Version')
+
+if [-z "$VERSION"]; then
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh]: Fixed

File: aws/eks/sentinel.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/sentinel.tf:9]: Nope not neede

File: aws/eks/vpn.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/vpn.tf:97]: It does and it does

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+role_arn="arn:aws:iam::$ACCOUNT_ID:role/lambda-list-layers"
+session_name="${2:-session-$(date +%s)}"
+
+temp_creds=$(aws sts assume-role \
+      --role-arn "$role_arn" \
+      --role-session-name "$session_name" \
+      --duration-seconds 3600 \
+      --output json)
+
+export AWS_ACCESS_KEY_ID=$(echo $temp_creds | jq -r .Credentials.AccessKeyId)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:13]: not neede

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+export AWS_SECRET_ACCESS_KEY=$(echo $temp_creds | jq -r .Credentials.SecretAccessKey)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:14]: not needed

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+export AWS_SESSION_TOKEN=$(echo $temp_creds | jq -r .Credentials.SessionToken)
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:15]: not neede

File: aws/eks/helper_scripts/getSentinelLayerVersion.sh [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+VERSION=$(aws lambda list-layer-versions \
+  --layer-name aws-sentinel-connector-layer \
+  --query "max_by(LayerVersions, &Version)" | \
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/helper_scripts/getSentinelLayerVersion.sh:19]: The query will only ever return one result, so this is not needed

File: aws/eks/sentinel.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/sentinel.tf:9]: Nope, not required.

File: aws/eks/sentinel.tf [hunk 14]
```diff
@@ -12,7 +19,7 @@
   function_name     = "sentinel-cloud-watch-forwarder"
   billing_tag_value = "notification-canada-ca-${var.env}"
 
-  layer_arn = "arn:aws:lambda:ca-central-1:${var.sentinel_sre_aws_account_id}:layer:aws-sentinel-connector-layer:${var.sentinel_layer_version}"
+  layer_arn = "arn:aws:lambda:ca-central-1:${var.sentinel_sre_aws_account_id}:layer:aws-sentinel-connector-layer:${data.external.get_sentinel_layer_version.result.version}"
```
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/sentinel.tf:22]: I mean it is

File: aws/eks/vpn.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (ben851) [reply aws/eks/vpn.tf:97]: It is and it does

File: aws/eks/vpn.tf [hunk 15]
```diff
@@ -94,8 +94,7 @@ resource "aws_acmpca_certificate" "client_vpn" {
 
 data "external" "get_pca_arn" {
   # Get the PCA ARN
-  # Yes this is horrific. I'm sorry.
-  program = [var.env == "production" ? "../../../../../../../../scripts/getPCAARN.sh" : "../../../../../../../scripts/getPCAARN.sh"]
```
Comment (patheard) [inline aws/eks/vpn.tf:98]: 🥲

Comment (ben851) [reply aws/eks/vpn.tf:98]: I honestly would have kept it as a matter of pride - but it was breaking when I tried to pass multiple arguments.

**Review by P0NDER0SA — APPROVED** [review]
It's OK!

**Review by sastels — APPROVED** [review]
LGTM!

**Pull request merged**